systems/controllers/osc/operational_space_control_mbp.cc

```diff
@@ -0,0 +1,708 @@
+    }
+  }
+  // 5. Input constraint
+  if (with_input_constraints_) {
+    prog_->AddLinearConstraint(MatrixXd::Identity(n_u_, n_u_), u_min_, u_max_,
+                               u_);
+  }
+  // No joint position constraint in this implementation
+
+  // Add costs
+  // 1. input cost
+  if (W_input_.size() > 0) {
+    prog_->AddQuadraticCost(W_input_, VectorXd::Zero(n_u_), u_);
+  }
+  // 2. acceleration cost
+  if (W_joint_accel_.size() > 0) {
+    prog_->AddQuadraticCost(W_joint_accel_, VectorXd::Zero(n_v_), dv_);
+  }
+  // 3. Soft constraint cost
+  if (w_soft_constraint_ > 0) {
+    prog_->AddQuadraticCost(w_soft_constraint_ * MatrixXd::Identity(n_c_, n_c_),
+                            VectorXd::Zero(n_c_), epsilon_);
+  }
+  // 4. Tracking cost
+  for (unsigned int i = 0; i < tracking_data_vec_->size(); i++) {
+    tracking_cost_.push_back(prog_->AddQuadraticCost(MatrixXd::Zero(n_v_, n_v_),
+                                      VectorXd::Zero(n_v_), dv_)
+                                  .evaluator()
+                                  .get());
+  }
+}
+
+std::vector<bool> OperationalSpaceControlMBP::CalcActiveContactIndices(
+    int fsm_state) const {
+  std::vector<bool> active_contact_flags;
+  for (unsigned int i = 0; i < body_indices_.size(); i++) {
+    if (fsm_state_when_active_.empty()) {
+      active_contact_flags.push_back(true);
+    } else {
+      active_contact_flags.push_back(fsm_state_when_active_[i] == fsm_state);
+    }
+  }
+  return active_contact_flags;
+}
+
+drake::systems::EventStatus OperationalSpaceControlMBP::DiscreteVariableUpdate(
+    const drake::systems::Context<double>& context,
+    drake::systems::DiscreteValues<double>* discrete_state) const {
+  const BasicVector<double>* fsm_output =
+      (BasicVector<double>*)this->EvalVectorInput(context, fsm_port_);
+  VectorXd fsm_state = fsm_output->get_value();
+  const OutputVector<double>* robot_output =
+      (OutputVector<double>*)this->EvalVectorInput(context, state_port_);
+  double timestamp = robot_output->get_timestamp();
+
+  auto prev_fsm_state = discrete_state->get_mutable_vector(prev_fsm_state_idx_)
+                            .get_mutable_value();
+  if (fsm_state(0) != prev_fsm_state(0)) {
+    prev_fsm_state(0) = fsm_state(0);
+
+    discrete_state->get_mutable_vector(prev_event_time_idx_).get_mutable_value()
+        << timestamp;
+  }
+  return drake::systems::EventStatus::Succeeded();
+}
+
+VectorXd OperationalSpaceControlMBP::SolveQp(
+    const VectorXd& x_w_spr, const VectorXd& x_wo_spr,
+    const drake::systems::Context<double>& context, double t, int fsm_state,
+    double time_since_last_state_switch) const {
+  vector<bool> active_contact_flags = CalcActiveContactIndices(fsm_state);
+
+  std::unique_ptr<drake::systems::Context<double>> context_w_spr =
+      plant_w_spr_.CreateDefaultContext();
+  std::unique_ptr<drake::systems::Context<double>> context_wo_spr =
+      plant_wo_spr_.CreateDefaultContext();
```

Comment:
Making `context_w_spr` and `context_wo_spr` members of OSC can probably boost the speed a little bit (we don't need to allocate the memory for the context every time SolveQp is called)

Reply:
Done.